File: test-suite/cpp/guards_clean.cpp
#include <string>

// CatalogTotal checks the pointer before dereferencing (early exit).
int catalog_total(Envelope* envelope) {
    if (envelope == nullptr) {
        return 0;
    }
    return envelope->data.metrics.count.total;
}

// RegionName guards the client, then reads the chain.
std::string region_name(ShipClient* client) {
    if (client == NULL) {
        return "";
    }
    return client->conn.region.host.name;
}

// ZoomLevel uses the truthiness idiom.
int zoom_level(World* world) {
    if (!world) {
        return -1;
    }
    return world->map.zoom.level.max;
}

// TernaryFallback picks the guarded branch with ?:.
int ternary_fallback(Envelope* envelope) {
    return envelope != nullptr ? envelope->data.metrics.count.total : 0;
}

File: test-suite/swift/guards_clean.swift
import Foundation

// CatalogTotal binds the optional before reading it (guard let).
func catalogTotal(_ envelope: Envelope?) -> Int {
    guard let envelope = envelope, envelope.isValid else {
        return 0
    }
    return envelope.data.metrics.count.total
}

// HostName guards the client then reads the chain.
func hostName(_ client: ShipClient?) -> String {
    if client == nil {
        return ""
    }
    return client.conn.region.host.name
}

// RawLabel binds with if let.
func rawLabel(_ envelope: Envelope?) -> Int {
    if let env = envelope {
        return env.data.metrics.count.total
    }
    return 0
}

// SafeName uses optional chaining end-to-end, never force-unwrapping.
func safeName(_ envelope: Envelope?) -> Int {
    return envelope?.data?.metrics?.count?.total ?? 0
}
